nym-vpn/ui/src-tauri/src/commands/connection.rs

```diff
@@ -67,23 +70,37 @@ pub async fn connect(
     )
     .ok();
 
-    let app_config = config_store.lock().await.read().await.map_err(|e| {
-        CmdError::new(
-            CmdErrorSource::InternalError,
-            format!("failed to read app config: {}", e),
-        )
-    })?;
-    let mut vpn_config = create_vpn_config(&app_config);
-    {
-        let app_state = state.lock().await;
-        if let VpnMode::TwoHop = app_state.vpn_mode {
-            info!("2-hop mode enabled");
-            vpn_config.enable_two_hop = true;
-        } else {
-            info!("5-hop mode enabled");
+    let app_state = state.lock().await;
+
+    debug!(
+        "using hardcoded entry node location: {}",
+        DEFAULT_NODE_LOCATION
+    );
+    let entry_point = EntryPoint::Location(DEFAULT_NODE_LOCATION.into());
+    let exit_point = match app_state.exit_node_location {
+        Some(ref exit_node_location) => {
+            debug!("exit node location set, using: {}", exit_node_location.code);
+            ExitPoint::Location(exit_node_location.code.clone())
+        }
+        _ => {
+            debug!(
+                "exit node location not set, using default: {}",
+                DEFAULT_NODE_LOCATION
+            );
+            ExitPoint::Location(DEFAULT_NODE_LOCATION.into())
         }
+    };
+
+    let mut vpn_config = create_vpn_config(entry_point, exit_point);
+    if let VpnMode::TwoHop = app_state.vpn_mode {
+        info!("2-hop mode enabled");
+        vpn_config.enable_two_hop = true;
+    } else {
+        info!("5-hop mode enabled");
     }
     // vpn_config.disable_routing = true;
+    // !! release the mutex
+    drop(app_state);
```

Comment:
scope ;)

Reply:
yea but not so simple to scope those, let's stick with drop
a refactoring would be better here, extracting logic parts into dedicated function